### DISCUSSION TITLE
[QnA] 169페이지 하단. findFirst와 findAny는 언제 사용하나? 부분

### DISCUSSION THREAD

**Discussion by Youngju-Jang** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#49] [CHAPTER 5 스트림 활용 - Q&A] [answered]
- findFirst, findAny 메서드가 모두 필요한 이유는 병렬성때문이라고 한다.
    - 병렬성에서는 첫번째 요소를 찾기 어렵기때문.
- 요소의 반환 순서가 상관없다면 병렬스트림에서는 제약이 적은 findAny를 사용하라고 함

### 그렇다면, 
1. findFirst메서드의 경우 병렬스트림에서는 사용이 불가능하다는 뜻인가? 아니면 효율적이지 못하니 지양하라는 뜻인가?
2. 일반적으로 이제까지 봤던 stream() 메서드는 `병렬스트림`이 아닌가?
    - `parallelStream()` 이런식으로 호출한 Stream만 `병렬스트림` 인가?

**Comment by coalong** [comment, marked as answer]
1. 
 제가 책을 읽고 이해한  내용은 
`findFirst` 는 병렬스트림과 순차스트림 둘 다 사용할 수 있고, 또 둘 다 같은 결과(스트림의 첫 번째요소)를 반환합니다.  (순차와 병렬 둘 다 반환 순서 보장 O)
`findAny`는 병렬스트림과 순차스트림에서 무조건 결과를 찾는 즉시 그 값을 반환하기 때문에, 시간 비용이 효율적이라고 생각합니다. (순차와 병렬 둘 다 반환 순서 보장 X)

그래서 스트림의 논리적인 아이템 순서가 정해져있고 반환 순서가 중요한 경우는 비용이 더 들더라도 `findFirst` 를 사용하고(하지만 병렬 처리에서는 권장하지 X) , 그렇지 않은 경우에는 `findAny` 를 사용하는게 좋다. 
라고 이해했습니다.

2. stream 은 순차적 처리와 병렬 처리가 있고,
병렬 처리를 하기 위해서는 parallel(), parallelStream() 을 사용해야만 병렬 처리 된다고 이해했습니다.

**Comment by JoisFe** [comment]
![image](https://github.com/Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action/assets/90208100/06d47199-f271-48d0-a3e7-0bff29b56bf5)

병렬 처리를 하는 상황에서
사용하면 1 과 5 중에서 전체 스트림에서는 어떤 것이 첫번째 값인지 보장할 수가 없기에 만약 1을 찾는 것이 보장되어야 한다면 findFirst를 사용하는  것이라고 이해했습니다. 만약 첫번째 요소를 찾아야 하는게 아니라면 findAny를 사용하는게 아닌가? 싶습니다.
(위의 예시는 스트림의 모든 값을 더하는 연산을 병렬처리 하는 그림이라 적합하진 않지만 비슷하게 병렬 처리를 한 후 첫번째 값을 찾아야 하는 상황으로 이해하시면 좋을 것 같습니다.)
